README
(More details can be found on harmonySEQ's homepage: http://harmonyseq.wordpress.com)

To run harmonySEQ you need to have {gtkmm-2.4,glibmm-2.4,alsa,liblo} run-time libriries installed on your system (though you can disable OSC support and liblo dependency by passing --disable-osc to configure script).
To compile it, you will also need the developement headers of above libraries.

To compile:
    autoreconf --install
	./configure
	make
	
It is also reccomended to install harmonySEQ (f.e for full language support):

To install (as root):
	make install
	
To run:
	harmonySEQ
	
(or src/harmonySEQ if you didn't install it).


In case of any errors, please contact me, using the contact details on http://harmonyseq.wordpress.com
Any feedback is welcome.

Hope you'll find harmonySEQ usefull.

Rafał Cieślak, 2010-2012


CHANGELOG
unreleased:
	- Disabled i18n support to allow for cleaning up the build system.

harmonyseq 0.16.1:
	- Maintenance only release.
	- Ported user interface code to use GTK3 and gtkmm-3.0 / 3.24.
	- Fixed compatibility with modern C++ compilers.
	- Cleaned up unnecessary autotool files.
	- Fixed a major race condition related to opening files while sequencer diodes were operating.

harmonyseq 0.16 "ranpiollo"
	- Implemented an advanced piano-roll interface for sequencers.
	- Added a new type of sequencer - control sequencers, that are used to output MIDI event messages.
	- Added support for custom lengths and resolutions (this allows to play melodies in any time signature).
	- Added chart-like display for control sequencers, which makes it easy to edit their output.
	- Velocity is now applied to all notes separately.
	- German translation, thanks to Jan Fuchsman.
	- Basic Japanese translation, by Nel Pogorzelska.
	- New system for opening files, which allows to automatically convert older files.
	- Added possibility to hide chord's detailed settings.
	- Added some new chords.
	- Significant bug fixes (including much more efficient output routines, which result in MUCH less lag during playback).

harmonyseq 0.15 "arising"
	- Reorganized sequencer window, significantly improving the way a pattern is presented.
	- Merged the sequencer window and the events window into the main window.
	- Added colourful flashing playback indicators.
	- Added 4th note indicators.
	- Added PlayOnEdit feature.
	- Implemented OSC support.
	- Allowed reordering sequencers (this time seriously).
	- Actions that change chords overwrite octave only optionally (one can change chord without changing octave).
	- Added an action that transposes a sequencer by an octave.
	- New event that allows triggering events by an OSC message.
	- Increased tempo range, as some may need higher tempos.
	- Added a button that clears selected pattern.
	- Significantly improved data refreshing speed.
	- Many optimizations, in some cases harmonySEQ may use now up to 30% less CPU.
	- Improved support for unusual Gtk system themes.
	- Changed sequencer's default resolution to 16.
	- Many bugs fixed, and some minor GUI improvements. 

harmonyseq 0.14 "enlighted"
	- Moved GUI from buttons to an elegant toolbar and menu.
	- Reorganized chord logic and GUI.
	- Added a customizable metronome.
	- Added Preferences window for customizing harmonySEQ's settings - stored in ~/.harmonySEQ/harmonySEQ.conf.
	- Added many tooltips to make it easier to use harmonySEQ.
	- Added a pop-up menu for the sequencer list.
	- Added possibility to hide/show columns in the main window as a part of harmonySEQ's configuration.
	- Some major GUI improvements.
	- Fixed many bugs.
	- Added the README file.
	- Source code heavily cleaned-up.

harmonyseq 0.13 "polyphony"
	- Using checkboxes instead of sliders in seqencer GUI makes it now possible to produce polyponic voices using one sequencer
	- Added menubar to improve GUI
	- Added AboutBox
	- Included harmonySEQ icon and used it used widely
	- Basic shortcuts support
	- Added integration with desktop - menu entry, icon, MIME type support

harmonyseq 0.12
	- Added ability to bind multiple melodies to one sequencer
	- Added Play/Pause button
	- Added Sync action, that may be useful to synchronize harmonySEQ with an external application
	- Improved interaction with user
	- Recognizing wheter the file was modified since it was last saved
	- Added Play/pause action
	- Added action changing melodies in a specified sequencer
	- Reorganized Seq/on/off/toggle actions (merged into one action type), may cause problems when loading older version files.
	- Minor GUI improvements

harmonyseq 0.11
	- Advanced chord selection
	- Added action setting a sequencer's chord
	- Added action that plays a sequence once
	- Coloring rows in main window depending on sequencer state
	- Added guitar demo and self control demo
	- Minor GUI updates

harmonyseq 0.10
	- Events and actions implemented (as for now featuring only few events/actions types).
	- Midi events can be now passed-through (allows playing live on external controller)
	- Variable volume.
	- Ported to GNU Autotools.
	- Updated example files.
	- Opening files from command line.
	- Improved file opening/saving.
	- Many minor GUI fixes and improvements.

harmonyseq 0.9
	- Sequencers can be now safely dragged.
	- Sorting by several columns implemented (olny several, since it widens the columns).
	- Sequences may have different resolutions
	- User can choose sequence length

harmonyseq 0.8
    - File access rearanged
    - Sequencers may be now removed
    - ID's hidden in non-debug mode
    - Sequencers cab be now cloned
    - Flashing tempo indicator

harmonyseq 0.7
    - Saving and loading files

harmonyseq 0.6
    - Options affecting sequencers can be now set either in the main window or
        in the sequencer window.
    - Sequencer channel is displayed in the main window
    - Muted-unmuted changed to On-off

harmonyseq 0.5
    - Sequencers may be now added
    - Muting and name changing can be done through main window
    - Minor fixes


global.h
/*
    Copyright (C) 2010 Rafał Cieślak

    This file is part of harmonySEQ.

    HarmonySEQ is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    HarmonySEQ is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with HarmonySEQ.  If not, see <http://www.gnu.org/licenses/>.
*/


#ifndef GLOBAL_H
#define	GLOBAL_H
#include <gtkmm.h>
#include <config.h>
#include "EventsWindow.h"

#include <libintl.h>
#include <locale.h>
#include <fstream>
#include <vector>
#include <alsa/asoundlib.h>
#include <cstdlib>
#include <cstdio>
#include <string>
#include "gettext.h"
#include "messages.h"
using namespace std;
#define VERSION "0.10"
#define VERSION_A 0
#define VERSION_B 10
#define VERSION_C 0

#define ENDL "\n"
#define TICKS_PER_QUARTERNOTE 128
#define TICKS_PER_NOTE (4*TICKS_PER_QUARTERNOTE)
#define _(STRING) gettext(STRING)
#define DEFAULT_TEMPO 240.0

#define FLASH_INTERVAL 50
    //in ms
#define EVENTS_FLASH_TIMEOUT 200
    //in ms

#define SEQUENCE_DEFAULT_SIZE 8
#define NOTES_CONST_SIZE 6
#define DEFAULT_VOLUME 100

//the following may not work with gettext. careful!
#define SEQUENCER_DEFAULT_NAME _("sequencer")

#define RESOLUTIONS_NUM 7
#define RESOLUTIONS {1,2,4,8,16,32,64}
#define LENGTHS_NUM 7
#define LENGTHS {0.125,0.25,0.5,1.0,2.0,4.0,8.0}

#define FILE_GROUP_SYSTEM "System"
#define FILE_KEY_SYSTEM_TEMPO "tempo"
#define FILE_KEY_SYSTEM_MAINNOTE "mainnote"
#define FILE_KEY_SYSTEM_SEQ_NUM "sequencers_number"
#define FILE_KEY_SYSTEM_EVENTS_NUM "events_number"

#define FILE_GROUP_TEMPLATE_SEQ "Seq %d"
#define FILE_KEY_SEQ_RESOLUTION "resolution"
#define FILE_KEY_SEQ_LENGTH "length"
#define FILE_KEY_SEQ_ON "on"
#define FILE_KEY_SEQ_APPLY_MAIN_NOTE "apply_mn"
#define FILE_KEY_SEQ_CHANNEL "channel"
#define FILE_KEY_SEQ_VOLUME "volume"
#define FILE_KEY_SEQ_NAME "name"
#define FILE_KEY_SEQ_SEQUENCE "sequence"
#define FILE_KEY_SEQ_NOTES "notes"

#define FILE_GROUP_TEMPLATE_EVENT "Event %d"
#define FILE_KEY_EVENT_TYPE "type"
#define FILE_KEY_EVENT_ARG1 "arg1"
#define FILE_KEY_EVENT_ARG2 "arg2"
#define FILE_KEY_EVENT_ACTIONS_NUM "actions"

#define FILE_GROUP_TEMPLATE_EVENT_ACTION_TYPE "Action_%d_type"
#define FILE_GROUP_TEMPLATE_EVENT_ACTION_ARG1 "Action_%d_arg1"
#define FILE_GROUP_TEMPLATE_EVENT_ACTION_ARG2 "Action_%d_arg2"


#ifndef I_DO_NOT_WANT_EXTERNS_FROM_GLOBAL_H
class Sequencer;
class debug;
class error;
class MidiDriver;
class MainWindow;
class Event;
class EventsWindow;
extern vector<Sequencer *> sequencers;
extern int mainnote;
extern double tempo;
extern int ports_number;
extern int running;
extern int debugging;
extern debug* dbg;
extern error* err;
extern MidiDriver* midi;
extern MainWindow* mainwindow;
extern EventsWindow* eventswindow;
extern int passing_midi;
extern std::map<string, int> keymap_stoi;
extern std::map<int, string> keymap_itos;
extern vector<Event *> events;
#endif

bool Ask(Glib::ustring message, Glib::ustring secondary_message = "", bool lock_threads = false);
void Info(Glib::ustring message, Glib::ustring secondary_message = "", bool lock_threads = false);

#endif	/* GLOBAL_H */

